modules/git/object_format.go

```diff
@@ -5,97 +5,89 @@ package git
-	NewEmptyID() ObjectID
-
-	NewHasher() HasherInterface
+	// IsValid returns true if the input is a valid hash
+	IsValid(objIDStr string) bool
+	// NewIDFromString create ObjectID from object id string
+	NewIDFromString(objIDStr string) (ObjectID, error)
+	// ComputeHash computes and generate object id
+	ComputeHash(t ObjectType, content []byte) (ObjectID, error)
 }
 
-type Sha1ObjectFormat struct{}
+type Sha1ObjectFormatImpl struct{}
 
-func (*Sha1ObjectFormat) ID() ObjectFormatID { return Sha1 }
-func (*Sha1ObjectFormat) String() string     { return "sha1" }
-func (*Sha1ObjectFormat) Empty() ObjectID    { return &Sha1Hash{} }
-func (*Sha1ObjectFormat) EmptyTree() ObjectID {
-	return &Sha1Hash{
+var (
+	emptyObjectID = &Sha1Hash{}
+	emptyTree     = &Sha1Hash{
 		0x4b, 0x82, 0x5d, 0xc6, 0x42, 0xcb, 0x6e, 0xb9, 0xa0, 0x60,
 		0xe5, 0x4b, 0xf8, 0xd6, 0x92, 0x88, 0xfb, 0xee, 0x49, 0x04,
 	}
-}
-func (*Sha1ObjectFormat) FullLength() int { return 40 }
-func (*Sha1ObjectFormat) IsValid(input string) bool {
-	return sha1Pattern.MatchString(input)
-}
+)
 
-func (*Sha1ObjectFormat) MustID(b []byte) ObjectID {
-	var id Sha1Hash
-	copy(id[0:20], b)
-	return &id
+func (Sha1ObjectFormatImpl) Name() string { return "sha1" }
+func (Sha1ObjectFormatImpl) EmptyObjectID() ObjectID {
+	return emptyObjectID
 }
 
-func (h *Sha1ObjectFormat) MustIDFromString(s string) ObjectID {
-	return MustIDFromString(h, s)
+func (Sha1ObjectFormatImpl) EmptyTree() ObjectID {
+	return emptyTree
 }
-
-func (h *Sha1ObjectFormat) NewID(b []byte) (ObjectID, error) {
-	return IDFromRaw(h, b)
+func (Sha1ObjectFormatImpl) FullLength() int { return 40 }
+func (Sha1ObjectFormatImpl) IsValid(objIDStr string) bool {
+	return sha1Pattern.MatchString(objIDStr)
 }
 
-func (h *Sha1ObjectFormat) NewIDFromString(s string) (ObjectID, error) {
-	return genericIDFromString(h, s)
+func (h Sha1ObjectFormatImpl) NewIDFromString(objIDStr string) (ObjectID, error) {
+	return genericIDFromString(h, objIDStr)
 }
 
-func (*Sha1ObjectFormat) NewEmptyID() ObjectID {
-	return NewSha1()
+// ComputeHash compute the hash for a given ObjectType and content
+func (h Sha1ObjectFormatImpl) ComputeHash(t ObjectType, content []byte) (ObjectID, error) {
+	hasher := sha1.New()
+	_, _ = hasher.Write(t.Bytes())
+	_, _ = hasher.Write([]byte(" "))
+	_, _ = hasher.Write([]byte(strconv.FormatInt(int64(len(content)), 10)))
+	_, _ = hasher.Write([]byte{0})
+
+	var sha1 Sha1Hash
+	copy(sha1[:], hasher.Sum(nil))
+	return &sha1, nil
 }
 
-func (h *Sha1ObjectFormat) NewHasher() HasherInterface {
-	return &Sha1Hasher{sha1.New()}
+var Sha1ObjectFormat ObjectFormat = Sha1ObjectFormatImpl{}
+
+var SupportedObjectFormats = []ObjectFormat{
+	Sha1ObjectFormat,
+	// TODO: add sha256
 }
 
-func ObjectFormatFromID(id ObjectFormatID) ObjectFormat {
-	switch id {
-	case Sha1:
-		return &Sha1ObjectFormat{}
+func ObjectFormatFromName(name string) ObjectFormat {
+	if name == "" {
+		return Sha1ObjectFormat
```

Comment:
I understand the sentiment here, but this could cause mysterious errors down the road in situations where ObjectFormatName is lost somehow due to another bug. It will continue to work as nothing is wrong for Sha1 repositories and the rest will get mysterious bugs.

To make Gitea equally robust for all objectFormats, it would be better if this function returned `nil` and caused a panic. Git defaults to SHA1 only in `git init` and everything later is repository dependent.

This is my only concern with the changes in this PR.

Reply:
resolved